Allow gRPC service registration through extensible plugins

finnegancarroll opened the pull request:
### Description
Adds a `GrpcServiceFactory` interface as an extension point for other plugins to implement. `GrpcServiceFactory` provides an interface for wrapping a predefined `io.grpc.BindableService` which can be registered on the gRPC transport. Enabling plugins to provide their own service definitions opens the door for supporting pre-defined service schemas within the gRPC transport (such as [Flight](https://github.com/apache/arrow/tree/main/format) or [OTEL](https://github.com/open-telemetry/opentelemetry-proto)) proto types, as well as custom hand rolled services which do not mirror the REST API and are not tracked by the API spec.

### Related Issues
Resolves #19025

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

finnegancarroll commented:
Hi @andrross @karenyrx can you take a look at this approach?
Intentionally not including `GrpcServiceFactory` in the SPI as `BindableService` needs `io.grpc:grpc-core`.

karenyrx commented on modules/transport-grpc/src/main/java/org/opensearch/transport/grpc/GrpcServiceFactory.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Extension point for plugins to add a BindableService to the grpc-transport.
+ */
+public interface GrpcServiceFactory {
```
Would this interface be a better fit under the transport-grpc SPI (`modules/transport-grpc/spi/src/main/java/org/opensearch/transport/grpc/spi/`) ?

karenyrx commented on modules/transport-grpc/src/main/java/org/opensearch/transport/grpc/GrpcServiceFactory.java:
```diff
@@ -0,0 +1,31 @@
+     * @param client OpenSearch client for use in gRPC service definition.
+     * @return GrpcServiceFactory for chaining.
+     */
+    GrpcServiceFactory initClient(Client client);
```
I wonder if plugins will also need access to other things other than just client, e.g. registering new settings, or thread pools etc.  

To support that, we could potentially expose more fields like: 
```
    GrpcServiceFactory init(GrpcServiceContext context);
```
```
public class GrpcServiceContext {
    private final Client client;
    private final Settings settings;
    private final ThreadPool threadPool;
    // ... other dependencies as needed
}
```

karenyrx commented on modules/transport-grpc/src/main/java/org/opensearch/transport/grpc/GrpcPlugin.java:
```diff
@@ -103,6 +102,13 @@ public void loadExtensions(ExtensiblePlugin.ExtensionLoader loader) {
+        // Load discovered gRPC service factories
+        List<GrpcServiceFactory> services = loader.loadExtensions(GrpcServiceFactory.class);
+        if (services != null) {
+            servicesFactory.addAll(services);
```
We could also add some validation before registering the services, for example to detect:
1. if two plugins register services with the same gRPC service name which could cause a runtime conflict
2. if service names are following the incorrect service naming convention (not following kebab naming, or not dot separated, etc)

Additionally, it might be useful to have some logs to know 
1. which plugins were successfully/unsuccessfully registered
2. which plugins registered which services 


A way this could potentially be done is by adding some additional methods in the `GrpcServiceFactory` interface like: 
- `getGrpcServiceName()` - Actual gRPC service name for conflict detection
- `getPluginName()`- for plugin ownership identification, etc

karenyrx commented on modules/transport-grpc/src/main/java/org/opensearch/transport/grpc/GrpcServiceFactory.java:
```diff
@@ -0,0 +1,31 @@
+     * Build gRPC service.
+     * @return BindableService.
+     */
+    BindableService build();
```
Do we want to build this interface with the intent of supporting multiple gRPC services per plugin in the future? If so, we could refactor all the methods to return a `List`/`Set` of things, instead of just a singular object, like changing:
* `BindableService build()` -> `List<BindableService> buildServices()`
* `Stirng getGrpcServiceName()` ->  `Set<String> getGrpcServiceNames()`

finnegancarroll commented:
> @finnegancarroll I see you created a second SPI. Isn't BindableService a part of [grpc-api](https://mvnrepository.com/artifact/io.grpc/grpc-api/1.75.0), which only has dependencies on findbugs and errorprone annotations? That seems pretty lightweight and would be okay in a single SPI. What am I missing?

Hi @andrross, the main reason to separate this SPI was the guava runtime dependency. Guava is runtime only in the main module as well and plugins such as KNN which only want the protobuf definitions run into version and subsequent runtime issues within the plugin test infrastructure. Here is the workaround for the original KNN PR:
https://github.com/opensearch-project/k-NN/pull/2817/files#diff-49a96e7eea8a94af862798a45174e6ac43eb4f8b4bd40759b5da63ba31ec3ef7R343-R351

finnegancarroll commented:
> @finnegancarroll Does the new SPI really need a Guava runtime dependency? I guess I was thinking this new code would only require adding the grpc-api dependency, and that it could just be added to the existing SPI which would stay pretty lightweight. Does that not work?

Hi @andrross , doing some experimenting and guava can be excluded here and listed under `ignoreMissingClasses` in third party audit. The result would be guava is a unlisted runtime dependency however this seems relatively safe since it will probably never be the case that plugins using this interface will attempt to implement a `BindableService` from scratch and will almost certainly already have guava at runtime through dependencies they use to generate their `BindableService` implementations.

One other consideration I think is extending this SPI to include other `grpc-api` definitions in the future. What comes to mind is `ServerInterceptor` which is much more likely to be directly implemented, however i'm not sure this changes much as most plugins already have guava at runtime so it's difficult to imagine this being an issue still.

What do you think? Is it preferable to stick with the single SPI implementation and exclude guava or list it explicitly in a separate SPI?

andrross commented on modules/transport-grpc/transport-grpc-spi/build.gradle:
```diff
@@ -0,0 +1,42 @@
+
+dependencies {
+  api project(":server")
+  api "org.opensearch:protobufs:${versions.opensearchprotobufs}"
+  api "io.grpc:grpc-api:${versions.grpc}"
```
This looks broadly right to me. Both of these are really lightweight dependencies. That fact that guava is a runtime dependency of grpc-api is fine because the runtime in this context will be the transport-grpc module which has grpc-core (and guava and everything else needed to actually run a grpc server). The things that will take a dependency on this SPI just need the minimal grpc API classes as well as the OpenSearch protobuf definitions.

finnegancarroll commented:
Per offline convo with @andrross the second SPI is not needed as guava which is a transitive runtime dependency of `grpc-api` will of course always be present since implemented services are executed within the gRPC transport module. Removed the duplicate SPI and updated the readme to include `GrpcServiceFactory` endpoint.

finnegancarroll commented:
```
* What went wrong:
Execution failed for task ':distribution:bwc:minor:buildBwcLinuxTar'.
> Building 3.3.0 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/minor/build/bwc/checkout-3.3/distribution/archives/linux-tar/build/distributions/opensearch-min-3.3.0-SNAPSHOT-linux-x64.tar.gz
```

Seems related to recent 3.3.1 minor version.

finnegancarroll commented:
```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distribution:bwc:bugfix:buildBwcLinuxTar'.
> Building 3.3.1 didn't generate expected file /var/jenkins/workspace/gradle-check/search/distribution/bwc/bugfix/build/bwc/checkout-3.3/distribution/archives/linux-tar/build/distributions/opensearch-min-3.3.1-SNAPSHOT-linux-x64.tar.gz
```

Still seeing build failures. Up to date with main.
